fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1179,6 +1179,70 @@ void determineCommittedTransactions(CommitBatchContext* self) {
 	}
 }
 
+inline int64_t extractTenantIdFromStringRef(StringRef s) {
+	if (s.size() < TenantAPI::PREFIX_SIZE) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	// Parse mutation key to determine mutation encryption domain
+	StringRef prefix = s.substr(0, TenantAPI::PREFIX_SIZE);
+	return TenantAPI::prefixToId(prefix, EnforceValidTenantId::False);
+}
+
+int64_t extractTenantIdFromSingleKeyMutation(MutationRef m) {
+	ASSERT_WE_THINK(!isSystemKey(m.param1));
+	ASSERT_WE_THINK(isSingleKeyMutation((MutationRef::Type)m.type));
+
+	return extractTenantIdFromStringRef(m.param1);
+}
+
+// Check a single-key mutation is associated with a valid tenant id
+// FIXME(xwang): handle clear range properly
+ErrorOr<Void> checkTenantModeAccess(MutationRef m) {
+	if (isSystemKey(m.param1))
+		return Void();
+
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		auto tenantId = extractTenantIdFromSingleKeyMutation(m);
+		// throw exception for invalid raw access
+		if (tenantId == TenantInfo::INVALID_TENANT) {
+			return illegal_tenant_access();
+		}
+	} else {
+		// For clear range, we allow raw access
+		ASSERT_EQ(m.type, MutationRef::Type::ClearRange);
+		auto beginTenantId = extractTenantIdFromStringRef(m.param1);
+		auto endTenantId = extractTenantIdFromStringRef(m.param2);
+		CODE_PROBE(!(beginTenantId == endTenantId && beginTenantId != TenantInfo::INVALID_TENANT),
+		           "Clear Range raw access or cross multiple tenants");
+	}
+	return Void();
+}
+
+// Return true if all tenant check pass. Otherwise, return false and send error back
+bool validTenantModeAccess(const CommitTransactionRequest& tr, ProxyCommitData* const pProxyCommitData) {
```

Comment:
Nit: I would name this something slightly more general (like `validateTenantAccess`) since it's doing more than just checking the tenant mode